### PR TITLE
Move 'var/val parameter may not be call-by-name' to error case class

### PR DESCRIPTION
**Pull request by ennru**
@felixmulder One more and I call it a day.

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/reporting/diagnostic/messages.scala [hunk 1]
```diff
@@ -1336,4 +1336,18 @@ object messages {
           |out the parameter list when extending it.
           |"""
   }
+
+  case class VarValParametersMayNotBeCallByName(name: Names.TermName, mutable: Boolean)(implicit ctx: Context)
+    extends Message(VarValParametersMayNotBeCallByNameID) {
+    val msg = s"${if (mutable) "`var'" else "`val'"} parameters may not be call-by-name"
+    val kind = "Syntax"
+    val explanation =
+      hl"""Parameters of classes and traits may no be call-by-name. In case you
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/reporting/diagnostic/messages.scala]: They can be as long as they don't have a getter e.g, this compiles:
```scala
class Foo(x: => Int)
```

File: compiler/src/dotty/tools/dotc/reporting/diagnostic/messages.scala [hunk 2]
```diff
@@ -1336,4 +1336,18 @@ object messages {
...
+          |want the parameter to be evaluated on demand, consider using a function
+          |and a lazy value in the class such as
+          |  ${s"class MyClass(${name}Tick: () => String) {"}
+          |  ${s"  lazy val $name = ${name}Tick"}
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/reporting/diagnostic/messages.scala]: But that's different from what a by-name would do, the closest thing would be to use a `def`, not a `lazy val`:
```class MyClass(fooFun: () => Int)) {
  def foo = fooFun()
```

Comment (smarter) [inline compiler/src/dotty/tools/dotc/reporting/diagnostic/messages.scala]: Or even:
```scala
class MyClass(myFoo: => Int) {
  def foo() = myFoo
}
```

Comment (ennru) [inline compiler/src/dotty/tools/dotc/reporting/diagnostic/messages.scala]: Ok, that's better. Will update.

File: compiler/src/dotty/tools/dotc/reporting/diagnostic/messages.scala (hunk 1 above)
Comment (ennru) [inline compiler/src/dotty/tools/dotc/reporting/diagnostic/messages.scala]: Yes, will be more specific.

**Pull request merged**